File: sekre-mobile/composeApp/src/commonMain/kotlin/org/sekre_mobile/com/data/mapper/DivisionMapper.kt
package org.sekre_mobile.com.data.mapper

import org.sekre_mobile.com.data.mapper.MapperUtils.parseTimestamp
import org.sekre_mobile.com.data.remote.dto.response.DivisionDto
import org.sekre_mobile.com.data.remote.dto.response.DivisionMemberDto
import org.sekre_mobile.com.domain.entity.Division
import org.sekre_mobile.com.domain.entity.DivisionMemberUser
import org.sekre_mobile.com.domain.entity.DivisionRole

/**
 * Division Mapper
 * Data layer - converts between DTOs and domain entities
 */
object DivisionMapper {

    /** Convert DivisionDto to Division entity */
    fun DivisionDto.toDomain(): Division {
        return Division(
            id = id,
            organizationId = organizationId,
            name = name,
            createdAt = parseTimestamp(createdAt),
            updatedAt = updatedAt?.let { parseTimestamp(it) } ?: parseTimestamp(createdAt)
        )
    }

    /** Convert DivisionMemberDto to DivisionMemberUser entity. */
    fun DivisionMemberDto.toDomain(): DivisionMemberUser {
        return DivisionMemberUser(
            id = user.id,
            email = user.email,
            fullName = user.fullName,
            divisionRole = parseDivisionRole(divisionRole),
        )
    }

    private fun parseDivisionRole(role: String): DivisionRole = when (role.uppercase()) {
        "HEAD" -> DivisionRole.HEAD
        else -> DivisionRole.STAFF
    }
}

File: sekre-mobile/src/data/mappers/division.mapper.ts
import type {
  Division,
  DivisionId,
  DivisionMember,
  DivisionDetail,
  DivisionPage,
} from '@core/domain/entities/Division';
import type {
  DivisionDTO,
  DivisionMemberDTO,
  DivisionDetailDTO,
  DivisionListResponseDTO,
  DivisionResponseDTO,
  DivisionDetailResponseDTO,
} from '@data/dto/division.dto';
import { mapPaginationMeta } from './pagination.mapper';

export const mapDivisionDTOToEntity = (dto: DivisionDTO): Division => ({
  id: dto.id as DivisionId,
  name: dto.name,
  description: null, // tidak ada di response
  memberCount: 0, // tidak ada di response
  createdAt: new Date(dto.created_at),
  updatedAt: new Date(dto.updated_at),
});

export const mapDivisionMemberDTOToEntity = (dto: DivisionMemberDTO): DivisionMember => ({
  userId: dto.user.id,
  fullName: dto.user.full_name,
  email: dto.user.email,
  role: dto.division_role,
  joinedAt: new Date(dto.joined_at),
});

export const mapDivisionDetailDTOToEntity = (dto: DivisionDetailDTO): DivisionDetail => ({
  ...mapDivisionDTOToEntity(dto.division),
  members: dto.members.map(mapDivisionMemberDTOToEntity),
});

export const mapDivisionListDTOToPage = (dto: DivisionListResponseDTO): DivisionPage => ({
  items: dto.data.data.map(mapDivisionDTOToEntity),
  meta: mapPaginationMeta(dto.data.pagination),
});

export const mapDivisionResponseDTOToEntity = (dto: DivisionResponseDTO): Division =>
  mapDivisionDTOToEntity(dto.data);

export const mapDivisionDetailResponseDTOToEntity = (
  dto: DivisionDetailResponseDTO,
): DivisionDetail => mapDivisionDetailDTOToEntity(dto.data);
